pkg/scheduler/plugins/coescheduling/core/gang_cache.go

```diff
@@ -0,0 +1,232 @@
+	pgClient   pgclientset.Interface
+}
+
+func NewGangCache(args *config.CoschedulingArgs, podLister listerv1.PodLister, pgLister pglister.PodGroupLister, client pgclientset.Interface) *GangCache {
+	return &GangCache{
+		gangItems:  make(map[string]*Gang),
+		lock:       new(sync.RWMutex),
+		pluginArgs: args,
+		podLister:  podLister,
+		pgLister:   pgLister,
+		pgClient:   client,
+	}
+}
+
+func (gangCache *GangCache) createGangToCache(gangId string) *Gang {
+	gangCache.lock.Lock()
+	defer gangCache.lock.Unlock()
+
+	if gang, exist := gangCache.gangItems[gangId]; exist {
+		return gang
+	}
+
+	gang := NewGang(gangId)
+	gangCache.gangItems[gangId] = gang
+	return gang
+}
+
+func (gangCache *GangCache) getGangFromCache(gangId string) *Gang {
+	gangCache.lock.RLock()
+	defer gangCache.lock.RUnlock()
+
+	return gangCache.gangItems[gangId]
+}
+
+func (gangCache *GangCache) deleteGangFromCache(gangId string) {
+	gangCache.lock.Lock()
+	defer gangCache.lock.Unlock()
+
+	delete(gangCache.gangItems, gangId)
+}
+
+func (gangCache *GangCache) RecoverGangCache() {
+	// recover podGroup First
+	pgLister := gangCache.pgLister
+	pgList, err := pgLister.List(labels.NewSelector())
+	if err != nil {
+		klog.Errorf("RecoverGangCache podGroupList List error: %v", err.Error())
+	}
+	for _, podGroup := range pgList {
+		gangCache.onPodGroupAdd(podGroup)
+
+	}
+	// recover gang from pod
+	podLister := gangCache.podLister
+	podsList, err := podLister.List(labels.NewSelector())
+	if err != nil {
+		klog.Errorf("RecoverGangCache podsList List error: %v", err.Error())
+	}
+	for _, pod := range podsList {
+		gangCache.onPodAdd(pod)
+	}
+
+}
+
+func (gangCache *GangCache) onPodAdd(obj interface{}) {
+	pod, ok := obj.(*v1.Pod)
+	if !ok {
+		return
+	}
+
+	gangName := GetGangNameByPod(pod)
+	if gangName == "" {
+		return
+	}
+	var gang *Gang
+	// the gang is created in either CRD way or Annotation way
+	if _, exist := pod.Labels[v1alpha1.PodGroupLabel]; exist {
+		gangName := pod.Labels[v1alpha1.PodGroupLabel]
+		gangNamespace := pod.Namespace
+		gangId := GetNamespaceSplicingName(gangNamespace, gangName)
+		gang = gangCache.getGangFromCache(gangId)
+		if gang == nil {
+			gang = gangCache.createGangToCache(gangId)
+			klog.Infof("Pod comes earlier than PodGroup,so we create gang by pod on add, gangName: %v", gangId)
+		}
+	} else {
+		gangNamespace := pod.Namespace
+		gangName := pod.Annotations[extension.AnnotationGangName]
+		gangId := GetNamespaceSplicingName(gangNamespace, gangName)
+		gang = gangCache.getGangFromCache(gangId)
+		if gang == nil {
+			gang = gangCache.createGangToCache(gangId)
+			klog.Infof("Create gang by pod on add, gangName: %v", gangId)
+		}
+		if !gang.HasGangInit {
+			gang.tryInitByPodConfig(pod, gangCache.pluginArgs)
+			pgFromAnnotation := getPgFromGangAndPod(gang, pod)
+			_, err := gangCache.pgClient.SchedulingV1alpha1().PodGroups(pod.Namespace).Create(context.TODO(), pgFromAnnotation, metav1.CreateOptions{})
+			if err != nil {
```

Comment:
add retry, you can find example in unified-scheduler like "(processor *GpuModelDetermineProcessor) UpdateForGangRequest"